content/en/platform/corda/5.0/developing-applications/getting-started/utxo-advanced-ledger-extensions/building-basic-contract-design.md

```diff
@@ -0,0 +1,237 @@
+        onVerifyUpdate(transaction);
+    }
+    
+    private void verifyDelete(UtxoLedgerTransaction transaction) {
+        // Verify base Delete constraints
+        // Then verify additional Delete constraints implemented by derived contracts
+        onVerifyDelete(transaction);
+    }
+}
+```
+
+Refactoring a contract as shown in the above example allows CorDapp implementors to derive from the contract, allowing additional constraints which will be verified in additional to the constraints specified by the base contract.
+
+{{< note >}}
+There are still some outstanding issues with this design, where this design approach no longer fits the design goals of the system being implemented.
+{{</ note >}}
+
+The problem really lies in the `verify` function; for example:
+
+```kotlin
+public final void verify(UtxoLedgerTransaction transaction) {
+    List<? extends ExampleContractCommand> commands = transaction
+            .getCommands(ExampleContractCommand.class);
+            
+    for (ExampleContractCommand command : commands) {
+        if (command instanceof Create) verifyCreate(transaction);
+        else if (command instanceof Update) verifyUpdate(transaction);
+        else if (command instanceof Delete) verifyDelete(transaction);
+        else throw new IllegalStateException("Unrecognised command type.");
+    }
+}
+```
+
+The `verify` function is marked final for security reasons, and therefore additional commands cannot be added to the contract. For example, the contract may wish to describe multiple ways to `Update` a state, or set of states. The contract only defines a single `Update` command: there can only be one mechanism to perform updates.
+
+The second problem lies in the commands themselves and their names. `Create`, `Update` and `Delete` are very ambiguous names, which may not make sense depending on the context of the contract being implemented.
+
+## Delegated Command Design
+
+In the contracts above, the commands are nothing more than marker classes; effectively they are cases in a switch statement, which allows the contract's `verify` function to delegate responsibility of specific contract constraints to other functions, such as `verifyCreate`, `verifyUpdate` and `verifyDelete`.
+
+We can implement the `verify` function on the command itself. Instead of being an empty marker class, this gives the command responsibility, as it becomes responsible for implementing its associated contract verification constraints.
+In this case, we define a `VerifiableCommand` interface with a `verify` function; for example:
+
+```kotlin
+public interface VerifiableCommand extends Command {
+    void verify(UtxoLedgerTransaction transaction);
+}
+```
+
+Now that we have a command which itself can implement contract verification constraints, we can use this as the basis for the `ExampleContractCommand` class. This needs to be a class rather than an interface, because we need to be in complete control of its implementations for security.
```

Comment:
```suggestion
Now that you have a command which itself can implement contract verification constraints, you can use this as the basis for the `ExampleContractCommand` class. This needs to be a class rather than an interface, because you need to be in complete control of its implementations for security.
```

Reply:
Done.